101-bundle-json.md

```diff
@@ -852,13 +852,131 @@ The fields are defined as follows:
   - `EXTENSION NAME`: a unique name for an extension. Names SHOULD follow the dotted name format described earlier in this section.
   - The value of the extension must be valid JSON, but is otherwise undefined.
 
-The usage of extensions is undefined. However, bundles SHOULD be installable by runtimes that do not understand the extensions.
+### Required Extensions
+
+Some extensions MAY be required by a bundle and operations involving that bundle MUST fail if the runtime does not support that bundle. In this case, a bundle author MUST use the `requiredExtensions` array to define those extensions that are necessary for installing the bundle. The `requiredExtensions` array should contain the `EXTENSION NAME` defined in the `custom` object for any extensions that are required for bundle operations.
+
+A runtime SHOULD check that it supports any required custom actions before performing any operation. However, bundles SHOULD be installable by runtimes that do not support extensions that are NOT identified in the `requiredExtensions` field.  
```

Comment:
Have we decided on SHOULD? For extensions, SHOULD implies that a tool could not bother to check and end up running malbehaved invocationImages -- in the sense that they were intended to be run along with a specific extension that the runtime does not support. This, I dislike, in that there's no statement here that would create a requirement to indicate to the user that it cannot support the bundle. 

If I am reading this correctly, I would urge MUST on our tools here. Even non-supporting tools would then give the user an opportunity to do something different. With SHOULD, we seem to be saying `meh`. Do I misunderstand?

Reply:
I'd very much like a statement SHOULD about the failure case for tooling. If a run time does not support any specified extension -- or extensions in general, should those different cases be called out? -- I'd very much like to emphasize that the proper way to handle this is to return as much information about the extensions not supported or about the extension feature itself so that users can have something to deal with as a result of the failure. This may be obvious, but asserting a SHOULD is a great emphasis for tool builders here.

Reply:
I agree with @squillace here, but I wonder why the text switches from talking about required extensions to required custom actions. How about the following based on @jeremyrickard's suggested text?
```suggestion
A runtime MUST check that it supports any required extensions before performing any operation. If the runtime does not support the required extension(s), it MAY proceed with the operation or fail, however it MUST notify the user that it does not support the required extension(s). Bundles SHOULD be installable by runtimes that do not support extensions that are NOT identified in the `requiredExtensions` field. 
```